_source/_posts/2022-02-10-dotnet-dependency-injection-comparison.md

```diff
@@ -0,0 +1,640 @@
+
+namespace OktaDependencyInjection.Services.Interfaces
+{
+    public interface IOktaService
+    {
+        Task<List<User>> GetUsersAsync();
+    }
+}
+```
+
+This interface defines a method for `GetUsersAsync` whose implementation will reach out to your Okta server using the Okta SDK.
+
+Add the implementation by adding a file called `OktaService` to the `Services` folder with the following code.
+
+```csharp
+using OktaDependencyInjection.Settings;
+using Microsoft.Extensions.Options;
+using Okta.Sdk;
+using Okta.Sdk.Configuration;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace OktaDependencyInjection.Services
+{
+    public class OktaService : Interfaces.IOktaService
+    {
+        private OktaClient _client;
+
+        public OktaService(IOptions<OktaConfiguration> options)
+        {
+            _client = new OktaClient(new OktaClientConfiguration()
+            {
+                OktaDomain = options.Value.Domain,
+                Token = options.Value.Token
+            });
+        }
+
+        public async Task<List<User>> GetUsersAsync()
+        {
+            var users = await _client.Users.ToListAsync();
+            return users.Select(r => r as User).ToList();
+        }
+    }
+}
+```
+
+Here you are using the [options pattern](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-5.0) which, like dependency injection, helps with the separation of concerns. The `options` object contains the configuration items from your `appsettings` file. It uses these values to create an `OktaClient`, which can then be called to get the users from your Okta account.
+
+The last service to add is the `RandomUpperAndLowerService`. Start by adding the `IRandomUpperAndLowerService.cs` file to the `Services/Interfaces` folder with the code below.
+
+```csharp
+namespace OktaDependencyInjection.Services.Interfaces
+{
+    public interface IRandomUpperAndLowerService
+    {
+        string RandomizeString(string input);
+    }
+}
+```
+
+Now add its implementation to the `Services` folder. Call this file `RandomUpperAndLowerService.cs` and add the code below to it.
+
+```csharp
+using System;
+using System.Collections.Generic;
+
+namespace OktaDependencyInjection.Services
+{
+    public class RandomUpperAndLowerService : Interfaces.IRandomUpperAndLowerService
+    {
+        public string RandomizeString(string input)
+        {
+            List<char> chars = new();
+            var rand = new Random();
+
+            foreach (var c in input.ToCharArray())
+            {
+                var next = rand.Next(0, 2);
+                chars.Add(next == 0 ? char.ToUpper(c) : char.ToLower(c));
+            }
+
+            return new string(chars.ToArray());
+        }
+    }
+}
+```
+
+While it's a silly function, this class will take a string and randomize if the text should be uppercase or lowercase. It will `pRodUCe STriNgS lIkE THiS`.  
```

Comment:
```suggestion
While it's a silly function, this class will take a string and randomize if the text should be uppercase or lowercase. It will `pRodUCe STriNgS lIkE THiS`.  
```
removes ambiguity